### PR TITLE
IRSA-3476_add_preview_of_pngs_to_data_tab

### PR DESCRIPTION
**Pull request by deannaji**
Jira: https://jira.ipac.caltech.edu/browse/IRSA-3476
K8s: https://dean-irsa-3476.irsakudev.ipac.caltech.edu/applications/sofia/
IFE Branch: https://github.com/IPAC-SW/irsa-ife/pull/116

For test:
1- Run a search ex: m17, radius 200.
2- Click on "HAWC+" tab
3- Click on a row with a .png instead of .fits
4- The "Preview" tab should display that image.

### CODE REVIEW

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import {useStoreConnector} from '../../ui/SimpleComponent';
+import {getActiveTableId, getTblInfoById, getCellValue} from '../../tables/TableUtil';
+
+export function PngViewer() {
+    const [tbl_id] = useStoreConnector(() => () => getActiveTableId());
```
Comment (robyww) [inline src/firefly/js/visualize/ui/PngViewer.jsx:6]: I have not tested this yet, but I don't think this should work, If is does we might have a problem with `getStoreConnector`.  As the user clicks row to row the `tbl_id` does not change so the component as it is written should not update.  You really need go get `tbl_id` and `highlightedRow`.

**Comment by ejoliet** [comment]
@deannaji if i search on HAWK+ only, not 'Any', the preview tab doesn't work.

Also, if you search and go back to search something else, you get a grey screen... i can show you.

File: src/firefly/js/visualize/ui/PngViewer.jsx (hunk 1 above)
Comment (loitly) [reply src/firefly/js/visualize/ui/PngViewer.jsx:6]: It's working, but I am guessing it's because it's rendering more than it needs to.

**Comment by deannaji** [comment]
@loitly , They will create a separate column for the PNG's and the 'Data' tab wont be showing them. So for this PR I will keep it open until the change is made. 
Thanks :)

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    const {tableMeta, tableModel} = getTblInfoById(tbl_id);
```
[the remark this answers is not in the document]
Comment (deannaji) [reply src/firefly/js/visualize/ui/PngViewer.jsx]: Fixed, thanks!

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const highlightedRow = tableModel? tableModel.highlightedRow: 0;
+    const PngSource = tableMeta.PngSource ? tableMeta.PngSource : '';
```
[the remark this answers is not in the document]
Comment (deannaji) [reply src/firefly/js/visualize/ui/PngViewer.jsx]: Fixed!

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    const png_url = getCellValue(tableModel, highlightedRow, PngSource);
+
+    if (png_url && png_url.endsWith('.png')) {
```
[the remark this answers is not in the document]
Comment (deannaji) [reply src/firefly/js/visualize/ui/PngViewer.jsx]: Done!

File: src/firefly/js/data/MetaConst.js [hunk 5]
```diff
@@ -96,6 +96,9 @@ export const MetaConst = {
     /** the column name with the url or filename of the image data */
     DATA_SOURCE : 'DataSource',
 
+    /** column name includes the link to the PNG image preview */
+    PNG_SOURCE : 'PngSource',
```
[the remark this answers is not in the document]
Comment (deannaji) [reply src/firefly/js/data/MetaConst.js]: Fixed!

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        return (
+            <div style={{
+                display: 'flex',
+                flexDirection: 'column',
+                background: '#c8c8c8',
+                width: '100%',
+                height: '100%'
+            }}>
+                <div style={{overflow: 'auto', display: 'flex', justifyContent: 'center', alignItem: 'center'}}>
+                    <img src={png_url} alt={png_url} style={{maxWidth: '100%', flexGrow: 0, flexShrink: 0}}/>
```
[the remark this answers is not in the document]
Comment (deannaji) [reply src/firefly/js/visualize/ui/PngViewer.jsx]: Fixed!

**Comment by ejoliet** [comment]
@deannaji i saw another issue, when the row is pink (proprietary data), the preview tab shouldn't be showing anything, same ,message as 'Data' tab: "You do not have access to this data."

File: src/firefly/js/visualize/ui/PngViewer.jsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    const [tbl_id] = useStoreConnector(() => () => getActiveTableId());
+    const {tableModel} = getTblInfoById(tbl_id);
+    if (!tableModel){
+        return null;
+    }
+    if(tableModel.isFetching){
+        return;
+    }
+    const highlightedRow = tableModel? tableModel.highlightedRow: 0;
+    const PngSource= getMetaEntry(tbl_id, 'ImagePreview', '');
```
Comment (robyww) [inline src/firefly/js/visualize/ui/PngViewer.jsx:15]: should be referred to by the constant
```
const PngSource= getMetaEntry(tbl_id, MetaConst.ImagePreview, '');
```

Comment (robyww) [reply src/firefly/js/visualize/ui/PngViewer.jsx:15]: variable name should not be uppercase unless the are classes, constants, or react component names
`pngSource`

Comment (deannaji) [reply src/firefly/js/visualize/ui/PngViewer.jsx:15]: oops, that was a mistake! I will fix it.

**Comment by deannaji** [comment]
Issues fixed and K8s build updated.

**Comment by ejoliet** [comment]
Comment about something to check for near future:
 
When i did a search on 'm17', only for 'HAWC+', i went to first row of the instrument tab.
I went to 'preview', what i see is grey background, i would expect 'No preview displayed' because it's a FITS but i imagined that this will change again when we will get the column for PNG. I would suggest to just make sure that the code handle properly when the value of cell from which it triggers/reacts it's correctly checking if there is or not a PNG.

**Comment by ejoliet** [comment]
Also, on a pink row highlight, i still see 'No preview available' in the 'preview' tab where it should say 'You do not have access to this data.'

**Comment by robyww** [comment]
> Also, on a pink row highlight, i still see 'No preview available' in the 'preview' tab where it should say 'You do not have access to this data.'

@deannaji you need to use the `TableUtil` function `hasRowAccess`

**Comment by robyww** [comment]
> When i did a search on 'm17', only for 'HAWC+', i went to first row of the instrument tab.
I went to 'preview', what i see is grey background, i would expect 'No preview displayed' because it's a FITS but i imagined that this will change again when we will get the column for PNG.

It is possible that the file just did not have time to load and fail. I have told @deannaji not to try to check, because there is not way to accurately check. It should just load and fail.  However, one we get the the final version there should not be a mix of png and fits files.

**Comment by deannaji** [comment]
Will look into those issues.

**Comment by robyww** [comment]
> is there a way we can add a loading visual to the PNG preview? 

Yes, we can do this. It is more work. You load and image into an offset screen element (see `WebUtil.js`,`loadImage`), while it is loading you show the loading.  When you get the loaded promise you create a canvas the same size as the image, and copy the image onto the canvas.

**Comment by ejoliet** [comment]
@deannaji a bug i found in the way the toolbar gets enabled or not depending on whether i'm on the PNG preview or not.
If i go to coverage tab, the image toolbar gets activated and enabled, that is ok. Then i go back to the preview tab, and the toolbar is still on when it shouldn't.
The toolbar should be in sync with the tab displayed. In case of PNG, because there is no other features we can act on a PNG, for now this toolbar buttons should all be disabled. 
Please, try to fix that before merging ;-).

Thanks.

**Comment by ejoliet** [comment]
(Same happens when i go to Data tab and back to preview tab)

**Comment by deannaji** [comment]
Cool, I will fix this issue. Thanks :)

**Comment by deannaji** [comment]
Fixed, thanks!

**Pull request merged**